m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -504,68 +462,108 @@ public void run() {
-                    // If we don't get any new metrics after waiting the specified time,
-                    // flush what we have so far.
-                    if (!pendingMetrics.isEmpty()) {
-                        flushMetricIteration(SizedMetric.FLUSH);
-                    }
                 }
             } catch (InterruptedException e) {
                 // Don't care if we get interrupted - the finally block will clean up
             } finally {
-                flushRemainingMetrics();
-
-                // Always arrive at phaser to prevent deadlock
-                phaser.arrive();
+                drainQueue();
+                flushBuffered();
+                // Count down shutdown latch to notify reporter
+                shutdownLatch.countDown();
             }
         }
 
-        private void flushMetricIteration(SizedMetric sizedMetric) {
-            Metric metric = sizedMetric.getMetric();
-
+        private void process(SizedMetric sizedMetric) {
             if (sizedMetric == SizedMetric.FLUSH) {
-                flush(pendingMetrics, commonTags);
+                flushBuffered();
                 return;
             }
 
             int size = sizedMetric.getSize();
-
-            if (metricsSize + size > freeBytes) {
-                flush(pendingMetrics, commonTags);
-                metricsSize = 0;
+            if (metricsSize + size > payloadCapacity || elapsedMaxDelaySinceLastFlush()) {
+                flushBuffered();
             }
 
-            pendingMetrics.add(metric);
+            Metric metric = sizedMetric.getMetric();
 
+            metricsBuffer.add(metric);
             metricsSize += size;
         }
 
-        private void flushRemainingMetrics() {
+        private boolean elapsedMaxDelaySinceLastFlush() {
+            return lastBufferFlushTimestamp.plus(maxBufferingDelay.toMillis(), ChronoUnit.MILLIS)
+                    .isBefore(Instant.now(clock));
+        }
+
+        private void drainQueue() {
             while (!metricQueue.isEmpty()) {
                 SizedMetric sizedMetric = metricQueue.remove();
-
                 // Don't care about metrics that came in after close
                 if (sizedMetric == SizedMetric.CLOSE) {
                     break;
                 }
 
-                flushMetricIteration(sizedMetric);
+                process(sizedMetric);
+            }
+        }
+
+        private void flushBuffered() {
+            if (metricsBuffer.isEmpty()) {
+                return;
+            }
+
+            MetricBatch metricBatch = new MetricBatch();
+            metricBatch.setCommonTags(commonTags);
+            metricBatch.setMetrics(metricsBuffer);
+
+            try {
+                client.emitMetricBatch(metricBatch);
+            } catch (TException tException) {
+                LOG.warn("Failed to flush metrics: " + tException.getMessage());
             }
 
-            flush(pendingMetrics, commonTags);
+            metricBatch.setMetrics(null);
```

Comment:
what's the point of setting this? it's a method local var so no risk of re-use

Reply:
Great question, this is a residual of previous method, simply didn't touch it

Reply:
Will remove it